runtime/jvmti/jvmtiThread.c

```diff
@@ -1248,22 +1271,141 @@ jvmtiGetCurrentThread(jvmtiEnv* env,
 
 #if JAVA_SPEC_VERSION >= 19
 jvmtiError JNICALL
-jvmtiSuspendAllVirtualThreads(jvmtiEnv* env,
+jvmtiSuspendAllVirtualThreads(jvmtiEnv *env,
 	jint except_count,
-	const jthread* except_list)
+	const jthread *except_list)
 {
-	assert(!"jvmtiSuspendAllVirtualThreads unimplemented");
-	return JVMTI_ERROR_NONE;
-}
+	J9JavaVM *vm = JAVAVM_FROM_ENV(env);
+	J9VMThread *currentThread = NULL;
+	jvmtiError rc = JVMTI_ERROR_NONE;
+
+	Trc_JVMTI_jvmtiSuspendAllVirtualThreads_Entry(env);
+
+	rc = getCurrentVMThread(vm, &currentThread);
+	if (JVMTI_ERROR_NONE == rc) {
+		jint i = 0;
+		UDATA currentThreadEverSuspended = FALSE;
+
+		omrthread_monitor_enter(vm->liveVirtualThreadListMutex);
+		vm->internalVMFunctions->internalEnterVMFromJNI(currentThread);
+
+		ENSURE_PHASE_LIVE(env);
+		ENSURE_CAPABILITY(env, can_suspend);
+		ENSURE_CAPABILITY(env, can_support_virtual_threads);
 
+		ENSURE_NON_NEGATIVE(except_count);
+		if ((NULL == except_list) && (0 != except_count)) {
+			JVMTI_ERROR(JVMTI_ERROR_NULL_POINTER);
+		}
+
+		for (i = 0; i < except_count; ++i) {
+			jthread thread = except_list[i];
+			if ((NULL == thread)
+			|| !IS_VIRTUAL_THREAD(currentThread, J9_JNI_UNWRAP_REFERENCE(thread))
+			) {
+				JVMTI_ERROR(JVMTI_ERROR_INVALID_THREAD);
+			}
+		}
+
+		/* Walk all virtual threads. */
+		if (NULL != vm->liveVirtualThreadList) {
+			j9object_t root = *(vm->liveVirtualThreadList);
+			/* Skip the root, which is a dummy virtual thread and global ref. */
+			j9object_t walkVirtualThread = J9OBJECT_OBJECT_LOAD(currentThread, root, vm->virtualThreadLinkNextOffset);
+			do {
+				BOOLEAN suspend = TRUE;
+				for (i = 0; i < except_count; ++i) {
+					if (walkVirtualThread == J9_JNI_UNWRAP_REFERENCE(except_list[i])) {
+						suspend = FALSE;
+						break;
+					}
+				}
+				if (suspend) {
+					UDATA currentThreadSuspended = 0;
```

Comment:
This should use `FALSE` instead of `0` for consistency.

Reply:
Updated.